module/abci.go

```diff
@@ -12,52 +13,57 @@ import (
 
 // BeginBlocker updates the validator set without applying updates.
 // Since this module depends on staking, that module will `ApplyAndReturnValidatorSetUpdates` from x/staking.
-func (am AppModule) BeginBlocker(ctx context.Context) error {
+func (am AppModule) EndBlocker(ctx context.Context) ([]abci.ValidatorUpdate, error) {
 	sdkCtx := sdk.UnwrapSDKContext(ctx)
 	defer telemetry.ModuleMeasureSince(poa.ModuleName, sdkCtx.BlockTime(), telemetry.MetricKeyBeginBlocker)
 
 	vals, err := am.keeper.GetStakingKeeper().GetAllValidators(ctx)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	valUpdates := []abci.ValidatorUpdate{}
+
 	for _, v := range vals {
 		switch v.GetStatus() {
 		case stakingtypes.Unbonding:
 			// if the validator is unbonding, force it to be unbonded. (H+1)
 			v.Status = stakingtypes.Unbonded
 			if err := am.keeper.GetStakingKeeper().SetValidator(ctx, v); err != nil {
-				return err
+				return nil, err
 			}
+			valUpdates = append(valUpdates, v.ABCIValidatorUpdateZero())
 
 		case stakingtypes.Unbonded:
 			// if the validator is unbonded (above case), delete the last validator power. (H+2)
 			valAddr, err := sdk.ValAddressFromBech32(v.OperatorAddress)
 			if err != nil {
-				return err
+				return nil, err
 			}
 
 			if err := am.keeper.GetStakingKeeper().DeleteLastValidatorPower(ctx, valAddr); err != nil {
-				return err
+				return nil, err
 			}
 
 		case stakingtypes.Unspecified, stakingtypes.Bonded:
+			valUpdates = append(valUpdates, v.ABCIValidatorUpdate(v.Tokens))
 			continue
 		}
 	}
 
 	if sdkCtx.BlockHeight() > 1 {
 		// non gentx messages reset the cached block powers for IBC validations.
 		if err := am.resetCachedTotalPower(ctx); err != nil {
-			return err
+			return nil, err
 		}
 
 		if err := am.resetAbsoluteBlockPower(ctx); err != nil {
-			return err
+			return nil, err
 		}
 	}
 
-	return nil
+	// TODO: updates here
+	return valUpdates, err
```

Comment:
Can we just pass through the same values for all vals, is this cached lower or will it always set? may need to cache here on our side instead